src/providers/v2/quote-provider.ts

```diff
@@ -75,14 +82,32 @@ export class V2QuoteProvider implements IV2QuoteProvider {
             let outputAmount = amount.wrapped;
 
             for (const pair of route.pairs) {
-              if (pair.token0.equals(outputAmount.currency) && pair.token0.sellFeeBps?.gt(BigNumber.from(0))) {
-                const outputAmountWithSellFeeBps = CurrencyAmount.fromRawAmount(pair.token0, outputAmount.quotient);
-                const [outputAmountNew] = pair.getOutputAmount(outputAmountWithSellFeeBps);
-                outputAmount = outputAmountNew;
-              } else if (pair.token1.equals(outputAmount.currency) && pair.token1.sellFeeBps?.gt(BigNumber.from(0))) {
-                const outputAmountWithSellFeeBps = CurrencyAmount.fromRawAmount(pair.token1, outputAmount.quotient);
-                const [outputAmountNew] = pair.getOutputAmount(outputAmountWithSellFeeBps);
-                outputAmount = outputAmountNew;
+              if (amount.wrapped.currency.sellFeeBps) {
+                // this should never happen, but just in case it happens,
+                // there is a bug in sor. We need to log this and investigate.
+                const error = new Error(`Sell fee bps should not exist on output amount
+                ${JSON.stringify(amount)} on amounts ${JSON.stringify(amounts)}
+                on routes ${JSON.stringify(routes)}`)
+
+                // artificially create error object and pass in log.error so that
+                // it also log the stack trace
+                log.error({ error }, 'Sell fee bps should not exist on output amount')
+                metric.putMetric('V2_QUOTE_PROVIDER_INCONSISTENT_SELL_FEE_BPS_VS_FEATURE_FLAG', 1, MetricLoggerUnit.Count)
+              }
+
+              if (providerConfig.enableFeeOnTransferFeeFetching) {
+                if (pair.token0.equals(outputAmount.currency) && pair.token0.sellFeeBps?.gt(BigNumber.from(0))) {
+                  const outputAmountWithSellFeeBps = CurrencyAmount.fromRawAmount(pair.token0, outputAmount.quotient);
+                  const [outputAmountNew] = pair.getOutputAmount(outputAmountWithSellFeeBps);
+                  outputAmount = outputAmountNew;
+                } else if (pair.token1.equals(outputAmount.currency) && pair.token1.sellFeeBps?.gt(BigNumber.from(0))) {
+                  const outputAmountWithSellFeeBps = CurrencyAmount.fromRawAmount(pair.token1, outputAmount.quotient);
+                  const [outputAmountNew] = pair.getOutputAmount(outputAmountWithSellFeeBps);
+                  outputAmount = outputAmountNew;
+                } else {
+                  const [outputAmountNew] = pair.getOutputAmount(outputAmount);
+                  outputAmount = outputAmountNew;
+                }
```

Comment:
Did something change in this block of code?

Reply:
I realized a buggy scenario, which will be common in prod, if we ship routing-api PR today. bcc from the descriptions:

_There's a bug in the FOT quote and routes calculation path, so that it would cause interface to double quote FOT tax immediately, if we ship https://github.com/Uniswap/routing-api/pull/362 today._

_Let's use BULLET => WETH as an example. Hypothetically interface will issue two requests:_

_BULLET => WETH exact in with enableFeeOnTransferFeeFetching = true. Then it will writes the routes DB with the cached routes with the pool reserves with the buyFeeBps and sellFeeBps . Interface issued this request as part of their QA/testing._
_Degen uses interface issue a BULLET => WETH exact in quote. Since interface has not enabled to pass in enableFeeOnTransferFeeFetching , routing-api will have useCachedRoutes = true, and read from cached routes with the pool reserves with the buyFeeBps and sellFeeBps . Then in sor, it will calculate the routes and quotes with the buyFeeBps and sellFeeBps ._
_The problem is in (2), interface has no awareness that the quote and routes are from cached routes, and they have the hardcoded FOT tax logic in prod today https://github.com/Uniswap/interface/blob/main/src/hooks/useSwapTaxes.ts. Since [smart-order-router](https://github.com/Uniswap/smart-order-router) also quoted with FOT tax in [quote-provider](https://github.com/Uniswap/smart-order-router/blob/main/src/providers/v2/quote-provider.ts#L74-L110), this is effectively double FOT taxation in prod, before we enable FOT backend support to 100%. And since cached routes hit rate is over +80%, this bug is considered a blocker before we ship https://github.com/Uniswap/routing-api/pull/362._

Reply:
oh got it got it previously we were gating this on the presence of sellFeeBps instead of the flag 👍